### ISSUE TITLE
proper support for mapgen-related metadata

### ISSUE THREAD

**Issue by zeuner** [minetest/minetest#7110]
This issue report is targeted at planning the necessary architectural changes for metadata-based handling of mapgen incompatibility with customizable lua code.

This is (of course, one may think) a topic to be discussed based on its technical merits. However, due to past experiences I must say that there seem to be social issues within the developer community which may hinder a technical discussion. So, I will tackle these below in the hope that they can be also resolved. As these will likely be boring for the regular (technically-oriented) contributors, I will focus on the technical issue first, hoping that the two developers who hindered past technical discussions will resist the temptation to get personal on this topic again, leaving the technical discussion to developers who are actually interested in improving Minetest architecture a bit.

Experienced developers know that, when developing new code, it is hard to support all the mapgens supported by the default game. Therefore, there have been discussions to provide metadata-based hints to the engine in order to prevent or discourage incompatible choices when setting up a server. There has even been a workaround merged into the core engine which inaccurately treats mapgen incompatibility as an attribute of the chosen subgame.

Compared to a technically more accurate metadata tagging of mapgen incompatibility, this leads to less than optimal behaviour when developing. For example, when reusing some of the mods from a subgame for developing a new subgame, or for standalone customization of a server (where possible), there is no way to tell which mod(s) introduced the mapgen incompatibility that was tagged on the subgame level.

Obviously, the technically more appropriate way to tag mapgen incompatibility in metadata would be within the mod metadata. It is obvious because a subgame does not introduce code (that may be incompatible with a mapgen) itself - it is a collection of mods - which means that it is technically impossible for a subgame to introduce mapgen incompatibilities without those having already been introduced by one of its mods. This would lead to improvements with respect to modularity and reusability since the engine could even automatically deduce mapgen incompatibility in some cases (when mods are just being recombined into a new subgame - a common first step when starting development on a new subgame).

There have, however, been (unverified) concerns by developers who claim that the necessary architectural changes for supporting metadata processing directly from the mods would be too big to be practical. My first impression is that this may actually not be true, but I would prefer some constructive criticism from someone who actually does regular work on the mod manager code.

In any case, I think that architectural changes, should they be necessary to support this kind of logic, might actually be warranted because an existing code design that makes quick and rather problematic workarounds (deciding by subgame about mapgen compatibility) than the appropriate solution is usually an indication for technological debt.

I hope that bringing this up may lead to some technical discussion on the pros and cons.

Now for the social issues perceived:

Ideally, this last part of the posting will be irrelevant. However, if it is not, my hope is that we can also resolve these, since I believe an open source project can only benefit from technical discussions being performed instead of muted.

When I decided to contribute to Minetest, I carefully studied the contributing guidelines, and I do think I followed them. Also, I think that the Minetest development can possibly benefit from my contributions. I do a lot of work integrating customized Minetest setups, mostly for education-related scenarios. While doing so, and also while assisting people with their own efforts in customizing Minetest, I came across different occasions where the engine acts surprising for someone not fully familiar with it. A lot of my later contributions and efforts are targeted an reducing these surprises, making it easier for new developers to actually get a rewarding experience out of customizing Minetest. I believe that this is actually something where Minetest can benefit from since there a many enthusiastic people out there with great ideas, but where the implementation gets blocked by engine quirks.

However, I encountered some community behaviour which I am not used to from open source projects. Open source projects usually have some kind of technically oriented sharing culture where people have learned to see different opinions as beneficial since the technical discussions can lead to overall improvement of insight and code quality. However, there have been two developers who are active on github, and who seem to find it hard to accept different points of view on technical matters. I do not wish to name them since I don't want to make this personal, and since I think they will probably recognize their behaviour themselves when reading about it. One developer comments on almost every issue I open, providing criticism with usually not too deep argumentation behind it. On a properly working open source community, this would not be a problem since the resulting technical discussion would easily qualify the more profound argumentation. However, another developer acts with what appears as some kind of "protective behaviour" with respect to the first developer, muting the discussion thread at the point where an actual technical discussion would take place.

I hope that this can be resolved, and that these developers will find other ways of expressing themselves rather than getting personal on me. Also, I believe that there is no justifiable reason for muting a technical discussion on Minetest changes. If my arguments would turn out as invalid, this would easily be found out when actually considering them on a technical basis.

**Comment by rubenwardy** [comment]
duplicate of #6806
